Implement Issue #3 - allow mods to override whether an ability can be unlocked

Iridar opened the pull request:
Closes #3

I took a somewhat unconventional approach of passing a non-localized name with the Tuple that can be used by listeners to see why NPSBD considers this ability locked. That way mods can override the NPBSD's decision only in specific cases, like the ability being locked because Training Center is not built, but preserve the original behavior if the ability is locked because soldier's rank is not high enough.

I don't recall doing something like this with the Highlander, but I think this approach is good and flexible. I tested and everything seems to be working. Still, a review is requested.

pledbrook commented on CommunityPromotionScreen/Src/NewPromotionScreenbyDefault/Classes/NPSBDP_UIArmory_PromotionHero.uc:
```diff
@@ -567,35 +569,65 @@ function bool CanPurchaseAbilityEx(int Rank, int Branch, name AbilityName, out s
+	Tuple.Data[2].kind = XComLWTVName;
+	Tuple.Data[2].n = nReasonLocked;
+
+	`XEVENTMGR.TriggerEvent('OverrideCanPurchaseAbility', Tuple, UnitState);
```
`'OverrideCanPurchaseAbility'` can just be `Tuple.Id` to ensure consistency between the event ID and the tuple ID.

Iridar replied:
Done.

pledbrook commented on CommunityPromotionScreen/Src/NewPromotionScreenbyDefault/Classes/NPSBDP_UIArmory_PromotionHero.uc:
```diff
@@ -567,35 +569,65 @@ function bool CanPurchaseAbilityEx(int Rank, int Branch, name AbilityName, out s
+	/// nReasonLocked is passed in the Tuple so mods can see *why* NPSBD's considers this ability cannot be unlocked
+	/// without having to check the strLocReasonLocked localized string.
+	/// nReasonLocked can take the following values: 
+	/// 'NoTrainingCenter', 'NoClassPerkPurchased', 'NotHighEnoughRank', 'NotEnoughAP', 'LacksPrerequisites'.
```
If we're going for CHL doc syntax, then this would be a bit easier to read as:

```suggestion
	///  - 'NoTrainingCenter'
	///  - 'NoClassPerkPurchased'
	///  - 'NotHighEnoughRank'
	///  - 'NotEnoughAP'
	///  - 'LacksPrerequisites'
```

I would also consider using backticks for all references to variable names in the rest of the doc comment, just so that they stand out as code. Minor preference, though.

pledbrook commented on CommunityPromotionScreen/Src/NewPromotionScreenbyDefault/Classes/NPSBDP_UIArmory_PromotionHero.uc:
```diff
@@ -567,35 +569,65 @@ function bool CanPurchaseAbilityEx(int Rank, int Branch, name AbilityName, out s
+	/// Ability will be locked if either strLocReasonLocked or nReasonLocked hold any value, so 
+	/// if a mod needs to allow this ability to be unlocked, its listener needs to "empty out" both of them.
+	///
+	/// EventID: OverrideCanPurchaseAbility,
```
CHL docs syntax puts event definitions inside:
```
    ```event
    ...
    ```
```

Iridar replied:
Yeah but that's just for Make Docs generator, we don't have that, so I assumed we don't need it.

pledbrook commented on CommunityPromotionScreen/Src/NewPromotionScreenbyDefault/Classes/NPSBDP_UIArmory_PromotionHero.uc:
```diff
@@ -567,35 +569,65 @@ function bool CanPurchaseAbilityEx(int Rank, int Branch, name AbilityName, out s
 	{
-		ReasonLockedText = ReasonLacksPrerequisites;
+		strLocReasonLocked = ReasonLacksPrerequisites;
+		nReasonLocked = 'LacksPrerequisites';
```
Perhaps these should be constants so that mod authors won't be tripped up by mistyping one of these names.

Iridar replied:
Mods wouldn't be able to access those constants without building against CPS, which I don't think is warranted for something like this. Otherwise I would have made it a `enum`. We can still keep them as constants locally, if you think it's warranted, but they would not be used anywhere outside this one function anyway, so I don't think there's point in doing that.

pledbrook replied:
Good point. I didn't think about that. Yeah, we can't make it a requirement to build against the promotion screen.

pledbrook commented on CommunityPromotionScreen/Src/NewPromotionScreenbyDefault/Classes/NPSBDP_UIArmory_PromotionHero.uc:
```diff
@@ -567,35 +569,65 @@ function bool CanPurchaseAbilityEx(int Rank, int Branch, name AbilityName, out s
+	/// nReasonLocked can take the following values: 
+	/// 'NoTrainingCenter', 'NoClassPerkPurchased', 'NotHighEnoughRank', 'NotEnoughAP', 'LacksPrerequisites'.
+	/// Ability will be locked if either strLocReasonLocked or nReasonLocked hold any value, so 
+	/// if a mod needs to allow this ability to be unlocked, its listener needs to "empty out" both of them.
```
I don't understand why `nReasonLocked` is relevant as an output. If a mod wants to override the reason given, they *have* to provide `strLocReasonUnlocked`. I can understand `nReasonLocked` as an `in` argument, but it seems confusing that the mod author has to provide values for both or empty them both. And if they forget to do one of them, then `CanPurchaseAbilityEx()` automatically returns `false`.

Iridar replied:
> I don't understand why nReasonLocked is relevant as an output.

So that a mod can lock an ability without providing a reason for it. If you don't think that behavior should be allowed, then I'm fine with making it only an `in` parameter.

pledbrook replied:
Hmmm. For user experience, I would be tempted to make it a requirement to provide a reason (nothing more frustrating to players than not knowing why they can't select something). On the other hand, I feel a bit queasy forcing something on mods that may not always be reasonable. I'm definitely conflicted on this now that you've mentioned it.

I think on balance it will be less contentious and less frustrating for modders if they can block an ability selection without providing a reason at this point. It still seems a bit confusing to have to clear these two variables, one of which is solely there (in its `out` persona) to indicate whether to block the ability selection or not.

To be honest, I prefer APIs with obvious behaviours for in and out parameters. In this case, I would simply add a boolean `out` parameter that is explicitly the sole determinant as to whether the ability selection should be blocked or not. `nReasonLocked` can then be a straightforward `in` parameter representing what the default reason is, while `strLocReasonLocked` can be set or not. Yes it's an extra "unnecessary" argument, but it improves clarity of the API, and that makes it well worth it in my view.

Iridar replied:
Agreed, I'll adjust implementation.

pledbrook reviewed (changes requested):
Seems pretty straightforward. The two things I would really like to see are constants for the built-in reason names and turning `nReasonLocked` into a pure `in` parameter rather than `inout`. Reasons given inline (pun not intended).

Iridar commented:
I should add `Rank` and `Branch` to this Tuple as well.

Iridar commented:
Per discussion with Peter, I have adjusted the event trigger so that there's a simple bool property in the Tuple that will be used to determine whether the ability can be purchased or not.
Docs are adjusted accordingly, and I also added Rank and Row into the Tuple.
